core/node/storage/pg_stream_store.go

```diff
@@ -141,7 +142,186 @@ func NewPostgresStreamStore(
+}
+
+// maintainSchemaLock periodically checks the connection that established the
+// lock on the schema and will attempt to establish a new connection and take
+// the lock again if the connection is lost. If the lock is lost and cannot be
+// re-established, the store will send an exit signal to shut down the node.
+// This is blocking and is intended to be launched as a go routine.
+func (s *PostgresStreamStore) maintainSchemaLock(
+	ctx context.Context,
+	conn *pgxpool.Conn,
+) {
+	log := dlog.FromCtx(ctx)
+	defer conn.Release()
+
+	lockId := s.computeLockIdFromSchema()
+
+	count := 0
+	for {
+		// Check for connection health with a ping. Also, maintain the connection in the
+		// case of idle timeouts.
+		err := conn.Ping(ctx)
+		count++
+
+		if count%100 == 0 {
+			log.Debug("DB Ping!", "error", err)
+		}
+
+		if err != nil {
+			// We expect cancellation only on node shutdown. In this case,
+			// do not send an error signal.
+			if errors.Is(err, context.Canceled) {
+				return
+			}
+
+			log.Warn("Error pinging pgx connection maintaining the session lock, closing connection", "error", err)
+
+			// Close the connection to encourage the db server to immediately clean up the
+			// session so we can go ahead and re-take the lock from a new session.
+			conn.Conn().Close(ctx)
+			// Fine to call multiple times.
+			conn.Release()
+
+			// Attempt to re-acquire a connection
+			conn, err = s.acquireConnection(ctx)
+
+			// Shutdown the node for non-cancellation errors
+			if errors.Is(err, context.Canceled) {
+				return
+			} else if err != nil {
+				err = AsRiverError(err, Err_RESOURCE_EXHAUSTED).
+					Message("Lost connection and unable to re-acquire a connection").
+					Func("maintainSchemaLock").
+					Tag("schema", s.schemaName).
+					Tag("lockId", lockId).
+					LogError(dlog.FromCtx(ctx))
+				s.exitSignal <- err
+				return
+			}
+
+			log.Info("maintainSchemaLock: reacquired connection, re-establishing session lock")
+			defer conn.Release()
+
+			// Attempt to re-establish the lock
+			var acquired bool
+			err := conn.QueryRow(
+				ctx,
+				"select pg_try_advisory_lock($1)",
+				lockId,
+			).Scan(&acquired)
+
+			// Shutdown the node for non-cancellation errors.
+			if errors.Is(err, context.Canceled) {
+				return
+			} else if err != nil {
+				err = AsRiverError(err, Err_RESOURCE_EXHAUSTED).
+					Message("Lost connection and unable to re-acquire schema lock").
+					Func("maintainSchemaLock").
+					Tag("schema", s.schemaName).
+					Tag("lockId", lockId).
+					LogError(dlog.FromCtx(ctx))
+				s.exitSignal <- err
+			}
+
+			if !acquired {
+				err = AsRiverError(fmt.Errorf("schema lock was not available"), Err_RESOURCE_EXHAUSTED).
+					Message("Lost connection and unable to re-acquire schema lock").
+					Func("maintainSchemaLock").
+					Tag("schema", s.schemaName).
+					Tag("lockId", lockId).
+					LogError(dlog.FromCtx(ctx))
+				s.exitSignal <- err
+			}
+
+			time.Sleep(1 * time.Second)
```

Comment:
should this be a select on context cancelation/time.After? there is helper function base.SleepWithContext to do that.